### ISSUE TITLE
Test gids and bam related scripts under the igbh-small dataset, the accuracy is abnormal

### ISSUE THREAD

**Issue by WWWzq-01** [jeongminpark417/GIDS#12]
I am using the script from the evaluation directory to assess both Bam and gids under the igbh-small dataset. Below is the script that I used for testing Bam and gids:
- run_pg_gen.sh
- write_data.sh
- run_GIDS_IGB.sh
- run_BaM_IGB.sh

`run_pg_gen.sh` is used to generate page rank of the nodes
```shell
CUDA_VISIBLE_DEVICES=0  python /home/wzq/GIDS/evaluation/page_rank_node_list_gen.py --path /home/wzq/datasets/igbh_datasets  --dataset_size small --uva_graph 1    --data IGB   --emb_size 1024  
```
`write_data.sh` is used to wirte IGBH-small datasets to the SSD
```shell
/home/wzq/GIDS/bam/build/bin/nvm-readwrite_stripe-bench --input /home/wzq/datasets/igbh_datasets/small/processed/paper/node_feat.npy --queue_depth 1024 --access_type 1 --num_queues 128 --threads 102400 --n_ctrls 1 --loffset $((1200000*4096))

/home/wzq/GIDS/bam/build/bin/nvm-readwrite_stripe-bench --input /home/wzq/datasets/igbh_datasets/small/processed/author/node_feat.npy  --queue_depth 1024 --access_type 1 --num_queues 128 --threads 102400 --n_ctrls 1 --loffset $((2200000*4096))

/home/wzq/GIDS/bam/build/bin/nvm-readwrite_stripe-bench --input /home/wzq/datasets/igbh_datasets/small/processed/fos/node_feat.npy --queue_depth 1024 --access_type 1 --num_queues 128 --threads 102400 --n_ctrls 1 --loffset $((4126066*4096)) 

/home/wzq/GIDS/bam/build/bin/nvm-readwrite_stripe-bench --input /home/wzq/datasets/igbh_datasets/small/processed/institute/node_feat.npy --queue_depth 1024 --access_type 1 --num_queues 128 --threads 102400 --n_ctrls 1 --loffset $((4316515 * 4096)) 
```
the offset setting is set as the number of nodes of paper,fos,author,institude 
<img width="501" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/b4226e5c-958d-4bd4-95b2-735feb9fd058">

`run_BaM_IGB.sh` is used to evalute the Bam
```shell
CUDA_VISIBLE_DEVICES=0  python /home/wzq/GIDS/evaluation/heterogeneous_train.py --dataset_size small --path /home/wzq/datasets/igbh_datasets/   --epochs 1 --log_every 1000 --uva_graph 1 --GIDS --batch_size 1024  --data IGB --model_type rsage --num_layers 3 --fan_out '10,5,5' --cache_size $((4*1024)) --num_ssd 1   --num_ele $((550*1000*1000*1024)) --page_size 4096 --emb_size 1024  --pin_file pr_igbh_samll.pt
```
the output of `run_Bam_IGB.sh` is shown as below:
<img width="936" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/3fe7605f-11c3-4e7b-aa50-938c54b5466b">

`run_GIDS_IGB.sh` is used to evalute the GIDS
```shell
CUDA_VISIBLE_DEVICES=0  python /home/wzq/GIDS/evaluation/heterogeneous_train.py --dataset_size small --path /home/wzq/datasets/igbh_datasets/   --epochs 1 --log_every 1000 --uva_graph 1 --GIDS --batch_size 1024  --data IGB --model_type rsage --num_layers 3 --fan_out '10,5,5' --cache_size $((4*1024)) --num_ssd 1   --num_ele $((550*1000*1000*1024)) --page_size 4096 --emb_size 1024 --window_buffer --wb_size 8   --cpu_buffer --cpu_buffer_percent 0.2   --accumulator --pin_file pr_igbh_samll.pt

```
the output of `run_GIDS_IGB.sh` is shown as below:
<img width="942" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/f594372d-d188-40db-80eb-e66d9dc4e7c8">

I have passed the offset for each node type and passed the offset dictionary to GIDS with heterograph_map flag.

<img width="438" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/c0af99df-94d0-44c4-8f54-796e19277345">
<img width="606" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/cd57a3e5-3ddc-4a6b-8689-1d712d49d631">

I use /dev/libnvm0 for heterogenous graph.

-----
Note: I add evaluation code to `heterogenous_train` based on `heterogeneous_train_baseline.py`:
```python
    # Evaluation

    model.eval()
    predictions = []
    labels = []
    with torch.no_grad():
        for _, _, blocks in test_dataloader:
            blocks = [block.to(device) for block in blocks]
            inputs = blocks[0].srcdata['feat']
     
            if(args.data == 'IGB'):
                labels.append(blocks[-1].dstdata['label']['paper'].cpu().numpy())
            elif(args.data == 'OGB'):
                out_label = torch.index_select(label_array, 0, b[1]).flatten()
                labels.append(out_label.numpy())
            predict = model(blocks, inputs).argmax(1).cpu().numpy()
            predictions.append(predict)

        predictions = np.concatenate(predictions)
        labels = np.concatenate(labels)
        test_acc = sklearn.metrics.accuracy_score(labels, predictions)*100
    print("Test Acc {:.2f}%".format(test_acc)) 
```
Could you tell me why this accuracy is so low?

**Comment by WWWzq-01** [comment]
Actually, I have also tested gids under the igb-small dataset, but the accuracy is low too.

**Comment by jeongminpark417** [comment]
The accuracy should be around 70%. It seems like data is not written in SSD correctly. Let me try to reproduce this problem from our side.

**Comment by WWWzq-01** [comment]
Yeah, I think the reason might be that the data was not written to the SSD correctly. But I am not sure how to verify this.
This is the output of the igb_small dataset write:
begin
<img width="783" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/0ebca4bc-0416-402c-9a3a-8b5f14e444bf">
end:
<img width="831" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/0d6f58a4-d193-4c2d-94d5-65bd1b7e40f8">
And this is the output of the run_Bam_igb.sh script:
<img width="932" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/b1af68e8-18ba-4476-9cdf-9e305849d6e6">

**Comment by WWWzq-01** [comment]
This is the output of the igbh_small dataset write:
paper:
<img width="936" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/cfff02a6-295b-4cf8-9d61-bf909d48b3c3">
<img width="849" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/2c5f55a5-f162-4154-bff6-62913d41581f">
author:
<img width="1323" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/c21814d0-3ae0-4197-bbfe-ed623960d5a9">
<img width="968" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/b116d459-1678-4bf8-8427-e7a09868419c">

**Comment by jeongminpark417** [comment]
The reason that it did not write the data into SSD correctly is that the data you are using is a npy file format instead of a bin file. You can use tensor_write.py to write npy file into SSDs. You can use the same offset as you showed in the comment when you write, but when you read during training, use the same --offset flag (e.g. $((1200000*4096))) to point the first element in the SSD that you want to start reading to minimize the size of the metadata.

I have pushed the code and verified it so please pull the code before you test.

Please let us know if you face any problems.

**Comment by WWWzq-01** [comment]
I have used `tensor_write.py` to write the npy file into SSDs,but the accuracy remains low.
I tested the igb_small dataset with the default offset (offset=0).
```shell
python /home/wzq/GIDS/evaluation/tensor_write.py --path /home/wzq/datasets/igbh_datasets/small/processed/paper/node_feat.npy  --cache_size $((4*1024)) --num_ssd 1   --num_ele $((4*1000*1000*1024)) --page_size 4096 --ssd_list 0
```
To evaluate Bam, I used run_BaM_IGB.sh:
```shell
CUDA_VISIBLE_DEVICES=0  python /home/wzq/GIDS/evaluation/homogenous_train.py --dataset_size small --path /home/wzq/datasets/igb_datasets/   --epochs 1 --log_every 1000 --uva_graph 1 --GIDS --batch_size 1024  --data IGB --model_type sage --num_layers 3 --fan_out '10,5,5' --cache_size $((4*1024)) --num_ssd 1   --num_ele $((4*1000*1000*1024)) --page_size 4096 --emb_size 1024  --pin_file pr_small.pt
```
And for evaluating GIDS, I used run_GIDS_IGB.sh:
```shell
CUDA_VISIBLE_DEVICES=0  python  /home/wzq/GIDS/evaluation/homogenous_train.py --dataset_size small --path /home/wzq/datasets/igb_datasets/   --epochs 1 --log_every 1000 --uva_graph 1 --GIDS --batch_size 1024  --data IGB --model_type sage --num_layers 3 --fan_out '10,5,5' --cache_size $((4*1024)) --num_ssd 1   --num_ele $((4*1000*1000*1024)) --page_size 4096 --emb_size 1024 --window_buffer --wb_size 8   --cpu_buffer --cpu_buffer_percent 0.2   --accumulator --pin_file pr_small.pt
```
To check if the data was written to the SSD correctly, I modified tensor_write.py to use `read_feature` after `write_feature` to read the feature that was just written to the SSD.
```python
    emb = np.load(args.path)
    emb_tensor = torch.tensor(emb).to(device)
    # GIDS_Loader.store_tensor(emb_tensor, 0)
    GIDS_Loader.store_tensor(emb_tensor, offset=args.offset)
    
    test_emb = GIDS_Loader.read_tensor(10,offset=args.offset)
    print("GIDS: ", test_emb)
```
The parameters used for tensor_write.py are shown below:
```shell
python /home/wzq/GIDS/evaluation/tensor_write.py --path /home/wzq/datasets/igbh_datasets/small/processed/paper/node_feat.npy  --cache_size $((4*1024)) --num_ssd 1   --num_ele $((4*1000*1000*1024)) --page_size 4096 --ssd_list 0
```
The output is:
<img width="352" alt="image" src="https://github.com/jeongminpark417/GIDS/assets/52373880/b45eb1e6-acc7-4c22-a14d-c26f618b814c">
As shown in the figure, the return value of read_tensor is None.

**Comment by jeongminpark417** [comment]
Did you pull the most recent push? If you did and still having this issue, can you try to use readwrite benchmark in bam repository to write the file into a SSD (just one SSD). I think there is 128B meta data for npy file, so you can run tail command to parse the first 128B then write the file data into SSD.

**Comment by WWWzq-01** [comment]
I have pulled the most recent push. I set --ioffset=128 to write the data, and now the accuracy is normal (about 70%!).Thanks for your reply!!!
